src/Extensions.Hosting.AsyncInitialization/Hosting/AsyncInitializationHostExtensions.cs

```diff
@@ -18,19 +18,68 @@ public static class AsyncInitializationHostExtensions
+
+            await rootInitializer.InitializeAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Performs teardown, by calling all registered async initializers that implement <see cref="IAsyncTeardown"/>
+        /// </summary>
+        /// <param name="host">The host.</param>
+        /// <param name="cancellationToken">Optionally propagates notifications that the operation should be cancelled</param>
+        /// <returns>The <see cref="Task"/> that represents the asynchronous operation.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when the host is null.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when the initialization service has not been registered.</exception>
+        /// <exception cref="OperationCanceledException">Thrown when the cancellationToken is cancelled.</exception>
+        private static async Task TeardownAsync(this IHost host, CancellationToken cancellationToken = default)
+        {
+            if (host == null) throw new ArgumentNullException(nameof(host));
+
+            await using var scope = host.Services.CreateAsyncScope();
+            var rootInitializer = scope.ServiceProvider.GetService<RootInitializer>()
+                ?? throw new InvalidOperationException("The async initialization service isn't registered, register it by calling AddAsyncInitialization() on the service collection or by adding an async initializer.");
+
+            await rootInitializer.TeardownAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Initializes and runs the application, by first calling all registered async initializers.
+        /// After the host terminates, any registered initializers that implement <see cref="IAsyncTeardown"/> are called to perform the teardown.
+        /// The <paramref name="host"/> instance is disposed of after running.
+        /// </summary>
+        /// <param name="host">The <see cref="IHost"/> to initialize and run.</param>
+        /// <param name="cancellationToken">Optionally propagates notifications that the operation should be cancelled</param>
+        /// <returns>The <see cref="Task"/> that represents the asynchronous operation.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when the host is null.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when the initialization service has not been registered.</exception>
+        /// <exception cref="OperationCanceledException">Thrown when the cancellationToken is cancelled.</exception>
+        public static async Task InitAndRunAsync(this IHost host, CancellationToken cancellationToken = default)
+        {
+            if (host == null) throw new ArgumentNullException(nameof(host));
+
+            await using (host as IAsyncDisposable)
```

Comment:
If `host` _doesn't_ implement `IAsyncDisposable`, it won't be disposed at all... I think you should use the same approach as in [`RunAsync`](https://github.com/dotnet/runtime/blob/b9db657a0e7789ce97f6cb6683401fd31639d67c/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs#L63), and check both interfaces in a finally block.